akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +996,174 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
+      lazy val detachedException = new StreamDetachedException()
+      while (iterator.hasNext) { iterator.next().tryFailure(detachedException) }
+      waitingForProcessing.clear()
+    }
+
+    private def onAsyncInput(event: T) = interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+
+    private def sendEvent(event: T, promise: Promise[Done]): Promise[Done] = {
+      onAsyncInput(event)
+      currentState.get() match {
+        case Completed ⇒ failPromiseOnComplete(promise)
+        case _         ⇒ promise
+      }
+    }
+
+    // external call
+    override def invokeWithFeedback(event: T): Future[Done] = {
+      val promise: Promise[Done] = Promise[Done]()
+      promise.future.andThen { case _ ⇒ waitingForProcessing.remove(promise) }(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
+      waitingForProcessing.add(promise)
+      invokeWithPromise(event, promise).future
+    }
+
+    private def invokeWithPromise(event: T, promise: Promise[Done]): Promise[Done] = {
+      currentState.get() match {
+        // not started yet
+        case list @ Pending(_) ⇒
+          if (!currentState.compareAndSet(list, Pending(Event(event, OptionVal(promise)) :: list.pendingEvents)))
+            invokeWithPromise(event, promise) // atomicity is failed - try again
+          else promise
+        // started - can just send message to stream
+        case Initialized ⇒ sendEvent(event, promise)
+        // initializing is in progress in another thread (initializing thread is managed by Akka)
+        case Initializing ⇒ if (!currentState.compareAndSet(Initializing, Pending(Event(event, OptionVal(promise)) :: Nil))) {
+          (currentState.get(): @unchecked) match {
+            case Pending(_)  ⇒ invokeWithPromise(event, promise) // atomicity is failed - try again
+            case Initialized ⇒ sendEvent(event, promise)
+          }
+        } else promise
+        // fail promise as stream is completed
+        case Completed ⇒ failPromiseOnComplete(promise)
+      }
+    }
+
+    private def failPromiseOnComplete(promise: Promise[Done]): Promise[Done] = {
+      waitingForProcessing.remove(promise)
+      promise.tryFailure(new StreamDetachedException())
+      promise
+    }
+
+    //external call
+    override def invoke(event: T): Unit = {
+      @tailrec
+      def internalInvoke(event: T): Unit = currentState.get() match {
+        // started - can just send message to stream
+        case Initialized       ⇒ onAsyncInput(event)
+        // not started yet
+        case list @ Pending(l) ⇒ if (!currentState.compareAndSet(list, Pending(Event(event, OptionVal.None) :: l))) internalInvoke(event)
+        // initializing is in progress in another thread (initializing thread is managed by akka)
+        case Initializing ⇒ if (!currentState.compareAndSet(Initializing, Pending(Event(event, OptionVal.None) :: Nil))) {
+          (currentState.get(): @unchecked) match {
```

Comment:
What's the benefit of not just recursing directly? Isn't it now missing matching for `Completed`?